src/java.desktop/share/classes/sun/awt/image/OffScreenImageSource.java

```diff
@@ -201,9 +201,12 @@ private void produce() {
                 }
             }
         } catch (NullPointerException e) {
-            e.printStackTrace();
+            // If theConsumer is null and we throw a NPE when interacting with it:
+            // That's OK. That is an expected use case that can happen when an
+            // ImageConsumer detaches itself from this ImageProducer mid-production.
 
             if (theConsumer != null) {
```

Comment:
Do we need here and a few lines above save the theConsumer to the local, then check to null, then call imageComplete?

Reply:
My understanding is:

This ticket focuses on the use case where an ImageConsumer detaches itself from this OffScreenImageSource mid-production by calling `removeConsumer(ImageConsumer)`. So in this situation: I'd argue no, the ImageConsumer should not get an `imageComplete(int)` notification. Because it opted to remove itself.

Does this answer your question, or are you considering a different use case?

(Note: `addConsumer` is synchronized, so it's *usually* reasonable to assume the field `OffScreeImageSource#theConsumer` acts *mostly* like a local variable. I could contrive a new failing test case if one ImageConsumer called `addConsumer(newDifferentConsumer)` while receiving a notification, but that's a separate discussion. It's beyond the scope of this ticket, and (to my knowledge) it's a contrived edge case that nobody has ever actually complained about in the real world. And it would fail with or without the changes in this PR...)

Reply:
> Does this answer your question, or are you considering a different use case?

Yes, if it is safe to assume that it is not possible to change the `theConsumer` by a different thread.